OCaml 5.4 support

The latest version of num fails to build when using OCaml 5.4:
```
#=== ERROR while compiling num.1.5-1 ==========================================#
# context     2.4.0~alpha2 | linux/x86_64 | ocaml-base-compiler.5.4.0~alpha1 | https://opam.ocaml.org#584630e7a7e27e3cf56158696a3fe94623a0cf4f
# path        ~/.opam/default/.opam-switch/build/num.1.5-1
# command     ~/.opam/opam-init/hooks/sandbox.sh build make PROFILE=release opam-modern
# exit-code   2
# env-file    ~/.opam/log/num-20663-eff40a.env
# output-file ~/.opam/log/num-20663-eff40a.out
### output ###
# [...]
# ocamlc -w +a-4-9-41-42-44-45-48  -bin-annot -g -safe-string -strict-sequence -strict-formats -I +compiler-libs -c num_top_printers.mli
# ocamlc -w +a-4-9-41-42-44-45-48  -bin-annot -g -safe-string -strict-sequence -strict-formats -I +compiler-libs -c num_top_printers.ml
# ocamlc -w +a-4-9-41-42-44-45-48  -bin-annot -g -safe-string -strict-sequence -strict-formats -I +compiler-libs -c num_top.mli
# ocamlc -w +a-4-9-41-42-44-45-48  -bin-annot -g -safe-string -strict-sequence -strict-formats -I +compiler-libs -c num_top.ml
# File "num_top.ml", line 28, characters 36-61:
# 28 |                               (Ldot(Lident "Num_top_printers", s))
#                                          ^^^^^^^^^^^^^^^^^^^^^^^^^
# Error: This expression should not be a constructor, the expected type is
#        Longident.t Location.loc
# make[1]: *** [Makefile:86: num_top.cmo] Error 2
# make[1]: Leaving directory '/home/opam-health-check/.opam/default/.opam-switch/build/num.1.5-1/src'
# make: *** [Makefile:2: all] Error 2
```

--- a/toplevel/num_top.ml
+++ b/toplevel/num_top.ml
@@ -1,20 +1,17 @@
-(* Taken from findlib.  Findlib is copyright 1999 by Gerd Stolpmann and
-  distributed under the terms given in file LICENSE-findlib in this
-  directory. *)
-
-(*
-  Printers for types defined in the "num" library. Meant to be used as printers
-  in the ocaml toplevel. See num_top.mli.
-
-  Copyright (C) 2003  Stefano Zacchiroli <zack@debian.org>
-
-  Released under the same terms as findlib.
-
-  Simplified implementation for OCaml 4 and up, Xavier.Leroy@inria.fr
-
+(* Taken from the Zarith library.
+   It is distributed under LGPL 2 licensing, with static linking exception.
+   See the LICENSE file included in the distribution.
+   Contributed by Christophe Troestler.
+   Adapted to Num by Xavier Leroy.
 *)
 
-open Longident
+open Printf
+
+let eval_string
+      ?(print_outcome = false) ?(err_formatter = Format.err_formatter) str =
+  let lexbuf = Lexing.from_string str in
+  let phrase = !Toploop.parse_toplevel_phrase lexbuf in
+  Toploop.execute_phrase print_outcome err_formatter phrase
 
 let printers = [
   "nat_printer";
@@ -23,9 +20,9 @@ let printers = [
   "num_printer";
 ]
 
-let install_num_printer s =
-  Topdirs.dir_install_printer Format.err_formatter
-                              (Ldot(Lident "Num_top_printers", s))
-
-let _ =
-  List.iter install_num_printer printers
+let () =
+  let ok =
+    List.for_all
+      (fun p -> eval_string(sprintf "#install_printer Num_top_printers.%s;;" p))
+      printers in
+  if not ok then Format.eprintf "Problem installing Num printers@."

